File: extern_c/cpp/cats.cpp
#include "cats.hpp"

cat make_cat(const char *name)
{
    cat c{name, true};
    std::cout << "[C++] library says\n";
    return c;
}

const char *cat_name(const cat *c)
{
    return c->name_;
}

void cat_feed(cat *c)
{
    c->is_hungry = false;
}

void cat_meow(const cat *c)
{
    if (c->is_hungry)
    {
        std::cout << c->name_ << " is hungry\n";
    }
    else
    {
        std::cout << c->name_ << " is sleepy\n";
    }
}

File: extern_c/zig/cats.zig
pub const cat = extern struct {
    name_: c_str,
    is_hungry: bool,
};
pub const c_str = [*:0]const u8;

pub extern fn make_cat(name: c_str) cat;
pub extern fn cat_name(c: ?*const cat) c_str;
pub extern fn cat_feed(c: ?*cat) void;
pub extern fn cat_meow(c: ?*const cat) void;

// ?* optional poiter, like std::optional<> (null[opt] or value.*)
